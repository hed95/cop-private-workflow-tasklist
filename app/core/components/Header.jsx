import PropTypes from 'prop-types';
import React from 'react';
import {bindActionCreators} from 'redux';
import {connect} from 'react-redux';
import { withRouter, Link, NavLink } from 'react-router-dom';

// local imports
import AppConstants from "../../common/AppConstants";
import secureLocalStorage from '../../common/security/SecureLocalStorage';
import SkipLink from './SkipLink';


export class Header extends React.Component {
  constructor(props) {
      super(props);
      this.secureLocalStorage = secureLocalStorage;
      this.logout = this.logout.bind(this);
      this.state = {
        navMobileOpen: false,
        navData: [
          {
            id: 'home',
            urlStem: AppConstants.DASHBOARD_PATH,
            text: 'Home',
          },
          {
            id: 'tasks',
            urlStem: AppConstants.YOUR_TASKS_PATH,
            text: 'Tasks',
          },
          {
            id: 'forms',
            urlStem: AppConstants.FORMS_PATH,
            text: 'Forms',
          },
          {
            id: 'reports',
            urlStem: AppConstants.REPORTS_PATH,
            text: 'Reports',
          },
          {
            id: 'cases',
            urlStem: AppConstants.CASES_PATH,
            text: 'Cases',
          },
          {
            id: 'profile',
            urlStem: AppConstants.MY_PROFILE_PATH,
            text: 'My profile',
          },
        ]
      }
  }

  handleLinkClick() {
    this.setState({ navMobileOpen: false })
  };

  toggleNavMobileOpen(event) {
    event.preventDefault();
    const stateChange = this.state.navMobileOpen = !this.state.navMobileOpen
    this.setState({ navMobileOpen: stateChange })
  }

  logout(event) {
      event.preventDefault();
      this.secureLocalStorage.removeAll();
      this.props.kc.logout();
  }


  render() {
      return (
        <header className="govuk-header " role="banner" data-module="header">
          <SkipLink />
          <div className="govuk-header__container govuk-width-container">
            <div className="govuk-header__content">
              <Link 
                id='serviceName'
                to={AppConstants.DASHBOARD_PATH}
                className="govuk-header__link govuk-header__link--service-name"
              >
                {AppConstants.APP_NAME}
              </Link>
              <button 
                type="button" 
                className={
                  this.state.navMobileOpen === true 
                  ? 'govuk-header__menu-button govuk-js-header-toggle govuk-header__menu-button--open' 
                  : 'govuk-header__menu-button govuk-js-header-toggle'
                }
                aria-controls="navigation" 
                aria-label="Show or hide Top Level Navigation"
                onClick={event => this.toggleNavMobileOpen(event)}
              >
                Menu
              </button>
              <nav>
                <ul 
                  id="navigation" 
                  className={
                    this.state.navMobileOpen === true 
                    ? "govuk-header__navigation govuk-header__navigation--open" 
                    : "govuk-header__navigation"
                  }
                  aria-label="Top Level Navigation"
                >
                  {(this.state.navData).map(elem => {
                    const activeState = elem.active === true ? 'govuk-header__navigation-item--active' : '';
                    return (
                      <li className={`govuk-header__navigation-item ${activeState}`} key={elem.urlStem}>
                        <NavLink 
                          to={elem.urlStem} 
                          className="govuk-header__link" 
                          onClick={() => this.handleLinkClick(elem.urlStem)}
                          // GDS sets active style on the <li> element, but NavLink only works on the <a> 
                          // So setting active color here to ensure it's applied on click
                          activeStyle={{ color: "#1d8feb" }}
                        >
                          {elem.text}
                        </NavLink>
                      </li>
                    );
                  })}
                  <li className="govuk-header__navigation-item">
                    <a 
                      id='support'
                      className="govuk-header__link" 
                      href={AppConstants.SUPPORT_PATH}
                      target="_blank"
                      rel="noopener noreferrer"
                    >
                      Support
                    </a>
                  </li>
                  <li className="govuk-header__navigation-item">
                    <a
                      id='logout'
                      className="govuk-header__link" 
                      onClick={this.logout}
                    >
                      Sign out
                    </a>
                  </li>
                </ul>
              </nav>
            </div>
          </div>
        </header>
      );
    }
  }

Header.propTypes = {
history: PropTypes.shape({
    push: PropTypes.func,
}).isRequired,
kc: PropTypes.shape({
    logout: PropTypes.func,
}).isRequired,
};

const mapDispatchToProps = dispatch => bindActionCreators({}, dispatch);

export default withRouter(connect(state => ({
kc: state.keycloak,
appConfig: state.appConfig,
}), mapDispatchToProps)(Header));
