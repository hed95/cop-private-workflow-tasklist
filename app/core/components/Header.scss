.govuk-header__content {
  display: flex;
  justify-content: space-between;
  width: 100% !important;
}
