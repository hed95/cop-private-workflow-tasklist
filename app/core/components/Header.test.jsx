import React from 'react';
import { createMemoryHistory } from 'history';
import { Header } from './Header';

describe('Header', () => {
  const props = {
    history: createMemoryHistory(),
    appConfig: {
      serviceDeskUrls: {
        support: 'test',
      },
    },
    kc: {
      logout: jest.fn(),
    },
    location: {
      pathname: '/dashboard'
    }
  }

  it('renders without crashing', () => {
    const wrapper = shallow(<Header {...props} />);
    expect(wrapper.exists()).toBe(true);
  });

  it('matches snapshot', () => {
    const wrapper = shallow(<Header {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

});
