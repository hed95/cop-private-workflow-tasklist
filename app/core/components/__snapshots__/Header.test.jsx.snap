// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header matches snapshot 1`] = `
<header
  className="govuk-header "
  data-module="header"
  role="banner"
>
  <SkipLink />
  <div
    className="govuk-header__container govuk-width-container"
  >
    <div
      className="govuk-header__content"
    >
      <Link
        className="govuk-header__link govuk-header__link--service-name"
        id="serviceName"
        to="/dashboard"
      >
        Central Operations Platform
      </Link>
      <button
        aria-controls="navigation"
        aria-label="Show or hide Top Level Navigation"
        className="govuk-header__menu-button govuk-js-header-toggle"
        onClick={[Function]}
        type="button"
      >
        Menu
      </button>
      <nav>
        <ul
          aria-label="Top Level Navigation"
          className="govuk-header__navigation"
          id="navigation"
        >
          <li
            className="govuk-header__navigation-item "
            key="/dashboard"
          >
            <Link
              className="govuk-header__link"
              onClick={[Function]}
              to="/dashboard"
            >
              Home
            </Link>
          </li>
          <li
            className="govuk-header__navigation-item "
            key="/your-tasks"
          >
            <Link
              className="govuk-header__link"
              onClick={[Function]}
              to="/your-tasks"
            >
              Tasks
            </Link>
          </li>
          <li
            className="govuk-header__navigation-item "
            key="/forms"
          >
            <Link
              className="govuk-header__link"
              onClick={[Function]}
              to="/forms"
            >
              Forms
            </Link>
          </li>
          <li
            className="govuk-header__navigation-item "
            key="/reports"
          >
            <Link
              className="govuk-header__link"
              onClick={[Function]}
              to="/reports"
            >
              Reports
            </Link>
          </li>
          <li
            className="govuk-header__navigation-item "
            key="/cases"
          >
            <Link
              className="govuk-header__link"
              onClick={[Function]}
              to="/cases"
            >
              Cases
            </Link>
          </li>
          <li
            className="govuk-header__navigation-item "
            key="submit-a-form/edit-your-profile"
          >
            <Link
              className="govuk-header__link"
              onClick={[Function]}
              to="submit-a-form/edit-your-profile"
            >
              My profile
            </Link>
          </li>
          <li
            className="govuk-header__navigation-item"
          >
            <a
              className="govuk-header__link"
              href="https://support.cop.homeoffice.gov.uk/servicedesk/customer/portal/3"
              id="support"
              rel="noopener noreferrer"
              target="_blank"
            >
              Support
            </a>
          </li>
          <li
            className="govuk-header__navigation-item"
          >
            <a
              className="govuk-header__link"
              id="logout"
              onClick={[Function]}
            >
              Sign out
            </a>
          </li>
        </ul>
      </nav>
    </div>
  </div>
</header>
`;
